Add auto-generated Configuration schema from Pydantic model

davelopez opened the pull request:
Follow up to #16514

I was manually creating the model when I realized we already have the `config_schema.yml`. It would be much easier to maintain if we could update the model whenever we introduce a new config parameter. As wisely suggested by Marius, I will approach this by first generating the initial basic Pydantic Model from the config schema, and then, the idea is to use the model as the single source of truth and generate the schema from it.

### TODO
- [x] Generate initial Pydantic model from current `config_schema.yml`
- [x] Manually curate the model:
  - [x] Include any missing fields that are serialized in the API but not in the YAML schema (computed fields), etc.
  - [x] Create different `views` of the Configuration model, initially:
    - **ComputedGalaxyConfig**: contains common Galaxy configuration options computed at runtime (not defined in the YAML config schema).
    - **UserExposableComputedGalaxyConfig** / **AdminExposableComputedGalaxyConfig**: mainly used for discriminating admin/non-admin models, but can be used to extend the common `ComputedGalaxyConfig` with admin-only options.
    - **AdminExposableGalaxyConfig**/**UserExposableGalaxyConfig**: contains (admin-only or common) configuration options that are exposed through the API.
    - **FullGalaxyConfig**: contains all the options in the YAML configuration file. This model will be used to generate the YAML schema. These options are not exposed to the API.
  - [x] Add stricter types using literal values instead of strings
  - [x] Include https://github.com/galaxyproject/galaxy/pull/16561 in model
  - [x] Mark deprecated options as deprecated.
  - [x] Take `by_host` dynamic config options into account #12328
  - [x] Annotate `path_resolves_to` options
  - [x] Annotate `reloadable` options
  - [x] Figure out why converter tests are failing
    - After some digging, the problem was caused by `use_remote_user` getting a string (email) instead of a boolean as a value assigned from the `single_user` option in some cases. I decided to fix it by casting to the bool value which is more explicit https://github.com/galaxyproject/galaxy/pull/16518/commits/c9a679429201b8af4d463476012ab8070a387174
  - [ ] Fix API schema linting with custom annotations (per_host, path_resolves_to, etc...) 
- [ ] Generate `config_schema.yml` from the `Full` model

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - TBD

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Add basic script to generate GalaxyConfigModel

From the configuration schema.

davelopez committed:
Add makefile commands

And invoke `config-api-schema` when building the config options and updating the client API schema.

davelopez committed:
WIP: Add basic generated GalaxyConfigModel

This model is not yet properly generated as it exposes everything.

mvdbeek commented:
> Opening PR to gather feedback on this approach.

This is amazing, I love it. How about we turn this around and make the pydantic model the source of truth from which we generate the other stuff ... it should be much more expressive than what we have currently, and the model_validators would let us for instance easily flag stuff that can't be combined.

davelopez commented:
Thanks so much! I will definitely explore that, it should be way cleaner that way :+1:

davelopez committed:
Enhance GalaxyConfigModel generation

- Fix some encoding issues
- Use Annotated for fields

davelopez committed:
Revert "Add makefile commands"

This reverts commit 554ddee55ac56f6ab41827fc4c86234c069fd555.

davelopez committed:
WIP: Split config options into models

davelopez committed:
Fix some type annotations

davelopez committed:
Fix single_user field type duality

davelopez committed:
Fix metadata_strategy type annotation

davelopez committed:
Make optional options, well... optional

davelopez committed:
Annotate job_config_file option as optional

davelopez committed:
Add bool as possible WatchToolOptions type

See test/integration/test_data_manager.py::TestDataManagerIntegration::test_data_manager_installation_table_reload

davelopez committed:
Fix typo in tests for `extended_celery` option

The documentation says that the option is called `extended_celery` but
the tests were using `celery_extended`. The logic will accept both
options, but the tests should be consistent with the documentation.

davelopez committed:
Validate configuration on startup

The validation logic needs to live in the manager class because otherwise the configuration package will need to depend on the schema, pydantic, etc.

davelopez committed:
Handle `panel_views` type duality

When validating the config file, is a list of objects but when serialized to the API is a dictionary.

davelopez committed:
Refactor model names + update client schema

davelopez committed:
Add missing valid values for simplified_workflow_run_ui_target_history

davelopez committed:
Annotate more non-required options as optional

davelopez committed:
Add tool_shed_urls to admin computed configuration

davelopez committed:
Remove unused delay_tool_initialization option

davelopez committed:
Annotate more non-required config params as optional

davelopez committed:
Annotate deprecated options

davelopez committed:
Update client API schema

davelopez committed:
Annotate per_host options

davelopez committed:
Annotate the rest of custom properties

davelopez committed:
Fix bool assignment for use_remote_user

The option `use_remote_user` was used an declared as bool in the codebase, but it was assigned the value of `single_user` in some cases and that option can be either a bool or an email address (str) causing the type validation to fail.

davelopez committed:
Add basic schema generation (for testing)

davelopez committed:
Avoid unwanted spacing in descriptions

This is somewhat controversial, but otherwise the generated schema will have weird spacing and unwanted line breaks in all descriptions.

mvdbeek commented:
Discussed this in the backend channel, we're all on board with having the pydantic models as the source of truth, and that if necessary, we can replace the tooling built on pykwalify.
